Duplicates and "False Positives"

Hi, I tried this awesome tool on https://github.com/skypjack/entt and had some troubles.
The tool reported
1. a lot of Duplicates and
2. most reported Mutants, when replicated by hand, did not survive the test suite (the same executable).

The aforementioned Project uses `GTest`. I speculate that `mull` tests each test-case separately?

Steps to replicate:
- use ubuntu 18.04 and install `mull` via provided repo: https://mull.readthedocs.io/en/latest/Installation.html#install-on-ubuntu
- install `clang-9` (might want to test different clang versions)
- build `entt` with `CC=clang-9 CXX=clang++-9 cmake .. -DENTT_BUILD_TESTING=ON  -DCMAKE_CXX_FLAGS="-fembed-bitcode" -DCMAKE_EXPORT_COMPILE_COMMANDS=ON` in `./build`
- setup mull junk removal etc (https://mull.readthedocs.io/en/latest/fmtlibTutorial.html#junk-mutations)

my `mull` command than looks like this:
```
mull-cxx \
	--test-framework=GoogleTest \
	-compdb-path compile_commands.json \
	-compilation-flags="\
	-isystem /usr/bin/../lib64/gcc/x86_64-linux-gnu/8/../../../../include/c++/8 \
	-isystem /usr/bin/../lib64/gcc/x86_64-linux-gnu/8/../../../../include/x86_64-linux-gnu/c++/8 \
	-isystem /usr/bin/../lib64/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/backward \
	-isystem /usr/local/include \
	-isystem /usr/lib/llvm-9/lib/clang/9.0.0/include \
	-isystem /usr/include/x86_64-linux-gnu \
	-isystem /usr/include" \
	-ld-search-path=/lib/x86_64-linux-gnu \
	-ld-search-path=/usr/lib/x86_64-linux-gnu \
	--exclude-path _deps \
	-mutators cxx_all \
	test/view
```

<details>
  <summary>example output for `test/view`</summary>
  <!-- have to be followed by an empty line! -->

```
[info] Extracting bitcode from executable (threads: 1)
       [################################] 1/1. Finished in 27ms
[info] Loading bitcode files (threads: 4)
       [################################] 4/4. Finished in 182ms
[info] Compiling instrumented code (threads: 4)
       [################################] 4/4. Finished in 10ms
[info] Loading dynamic libraries (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Searching tests (threads: 1)
       [################################] 1/1. Finished in 2ms
[info] Preparing original test run (threads: 1)
       [################################] 1/1. Finished in 906ms
[info] Running original tests (threads: 4)
       [################################] 28/28. Finished in 74ms
[info] Applying function filter: no debug info (threads: 4)
       [################################] 2395/2395. Finished in 11ms
[info] Applying function filter: file path (threads: 4)
       [################################] 2395/2395. Finished in 10ms
[info] Instruction selection (threads: 4)
       [################################] 2344/2344. Finished in 16ms
[info] Searching mutants across functions (threads: 4)
       [################################] 2344/2344. Finished in 39ms
[info] Applying filter: no debug info (threads: 4)
       [################################] 3010/3010. Finished in 10ms
[info] Applying filter: file path (threads: 4)
       [################################] 3010/3010. Finished in 11ms
[info] Applying filter: junk (threads: 4)
       [################################] 3010/3010. Finished in 1530ms
[info] Prepare mutations (threads: 1)
       [################################] 1/1. Finished in 0ms
[info] Cloning functions for mutation (threads: 4)
       [################################] 4/4. Finished in 1089ms
[info] Removing original functions (threads: 4)
       [################################] 4/4. Finished in 483ms
[info] Redirect mutated functions (threads: 4)
       [################################] 4/4. Finished in 11ms
[info] Applying mutations (threads: 1)
       [################################] 179/179. Finished in 10ms
[info] Compiling original code (threads: 4)
       [################################] 4/4. Finished in 10ms
[info] Running mutants (threads: 4)
       [################################] 179/179. Finished in 10556ms
[info] Survived mutants (69/179):
/home/green/workspace/github/entt/src/entt/entity/registry.hpp:131:25: warning: Survived: Replaced < with <= [cxx_lt_to_le]
        return (!(index < pools.size()) || !pools[index].pool) ? nullptr : static_cast<const storage_type<Component> *>(pools[index].pool.get());
                        ^
/home/green/workspace/github/entt/src/entt/entity/registry.hpp:370:21: warning: Survived: Replaced < with <= [cxx_lt_to_le]
        return (pos < entities.size() && entities[pos] == entity);
                    ^
/home/green/workspace/github/entt/src/entt/entity/storage.hpp:173:24: warning: Survived: Replaced 'a = b' with 'a = 42' [cxx_assign_const]
        instances[pos] = std::move(other);
                       ^
/home/green/workspace/github/entt/src/entt/entity/storage.hpp:172:14: warning: Survived: Replaced 'T a = b' with 'T a = 42' [cxx_init_const]
        auto other = std::move(instances.back());
             ^
/home/green/workspace/github/entt/src/entt/entity/sparse_set.hpp:153:45: warning: Survived: Replaced & with | [cxx_and_to_or]
        return size_type{(to_integral(entt) & traits_type::entity_mask) / entt_per_page};
                                            ^
/home/green/workspace/github/entt/src/entt/entity/sparse_set.hpp:153:73: warning: Survived: Replaced / with * [cxx_div_to_mul]
        return size_type{(to_integral(entt) & traits_type::entity_mask) / entt_per_page};
                                                                        ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:894:45: warning: Survived: Replaced --x with ++x [cxx_pre_dec_to_pre_inc]
                for(auto pos = size(); pos; --pos) {
                                            ^
/home/green/workspace/github/entt/src/entt/entity/registry.hpp:131:25: warning: Survived: Replaced < with <= [cxx_lt_to_le]
        return (!(index < pools.size()) || !pools[index].pool) ? nullptr : static_cast<const storage_type<Component> *>(pools[index].pool.get());
                        ^
/home/green/workspace/github/entt/src/entt/entity/registry.hpp:388:49: warning: Survived: Replaced >> with << [cxx_rshift_to_lshift]
        return version_type(to_integral(entity) >> traits_type::entity_shift);
                                                ^
/home/green/workspace/github/entt/src/entt/entity/registry.hpp:149:47: warning: Survived: Replaced & with | [cxx_and_to_or]
        const auto entt = to_integral(entity) & traits_type::entity_mask;
                                              ^
/home/green/workspace/github/entt/src/entt/entity/storage.hpp:173:24: warning: Survived: Replaced 'a = b' with 'a = 42' [cxx_assign_const]
        instances[pos] = std::move(other);
                       ^
/home/green/workspace/github/entt/src/entt/entity/storage.hpp:172:14: warning: Survived: Replaced 'T a = b' with 'T a = 42' [cxx_init_const]
        auto other = std::move(instances.back());
             ^
/home/green/workspace/github/entt/src/entt/entity/storage.hpp:154:45: warning: Survived: Replaced - with + [cxx_sub_to_add]
            const auto pos = size_type(index-1u);
                                            ^
/home/green/workspace/github/entt/src/entt/entity/registry.hpp:144:61: warning: Survived: Replaced & with | [cxx_and_to_or]
        available = entity_type{to_integral(entities[curr]) & traits_type::entity_mask};
                                                            ^
/home/green/workspace/github/entt/src/entt/entity/registry.hpp:144:19: warning: Survived: Replaced 'a = b' with 'a = 42' [cxx_assign_const]
        available = entity_type{to_integral(entities[curr]) & traits_type::entity_mask};
                  ^
/home/green/workspace/github/entt/test/entt/entity/view.cpp:315:15: warning: Survived: Replaced 'a = b' with 'a = 42' [cxx_assign_const]
        check = false;
              ^
/home/green/workspace/github/entt/test/entt/entity/view.cpp:329:15: warning: Survived: Replaced 'a = b' with 'a = 42' [cxx_assign_const]
        check = false;
              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with >= [cxx_lt_to_ge]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with <= [cxx_lt_to_le]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with <= [cxx_lt_to_le]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/sparse_set.hpp:513:23: warning: Survived: Replaced 'T a = b' with 'T a = 42' [cxx_init_const]
        for(size_type pos{}; pos < count; ++pos) {
                      ^
/home/green/workspace/github/entt/src/entt/entity/sparse_set.hpp:513:34: warning: Survived: Replaced < with >= [cxx_lt_to_ge]
        for(size_type pos{}; pos < count; ++pos) {
                                 ^
/home/green/workspace/github/entt/src/entt/entity/sparse_set.hpp:513:43: warning: Survived: Replaced ++x with --x [cxx_pre_inc_to_pre_dec]
        for(size_type pos{}; pos < count; ++pos) {
                                          ^
/home/green/workspace/github/entt/src/entt/entity/sparse_set.hpp:513:23: warning: Survived: Replaced 'T a = b' with 'T a = 42' [cxx_init_const]
        for(size_type pos{}; pos < count; ++pos) {
                      ^
/home/green/workspace/github/entt/src/entt/entity/sparse_set.hpp:513:34: warning: Survived: Replaced < with >= [cxx_lt_to_ge]
        for(size_type pos{}; pos < count; ++pos) {
                                 ^
/home/green/workspace/github/entt/src/entt/entity/sparse_set.hpp:513:43: warning: Survived: Replaced ++x with --x [cxx_pre_inc_to_pre_dec]
        for(size_type pos{}; pos < count; ++pos) {
                                          ^
/home/green/workspace/github/entt/test/entt/entity/view.cpp:581:20: warning: Survived: Replaced < with <= [cxx_lt_to_le]
        return lhs < rhs;
                   ^
/home/green/workspace/github/entt/test/entt/entity/view.cpp:589:20: warning: Survived: Replaced > with >= [cxx_gt_to_ge]
        return lhs > rhs;
                   ^
/home/green/workspace/github/entt/test/entt/entity/view.cpp:589:20: warning: Survived: Replaced > with <= [cxx_gt_to_le]
        return lhs > rhs;
                   ^
/home/green/workspace/github/entt/test/entt/entity/view.cpp:599:20: warning: Survived: Replaced < with <= [cxx_lt_to_le]
        return lhs < rhs;
                   ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with >= [cxx_lt_to_ge]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with <= [cxx_lt_to_le]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with >= [cxx_lt_to_ge]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with <= [cxx_lt_to_le]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:231:21: warning: Survived: Replaced 'T a = b' with 'T a = 42' [cxx_init_const]
        std::size_t pos{};
                    ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/storage.hpp:85:20: warning: Survived: Replaced --x with ++x [cxx_pre_dec_to_pre_inc]
            return --index, *this;
                   ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:233:159: warning: Survived: Replaced x++ with x-- [cxx_post_inc_to_post_dec]
        (static_cast<void>(std::get<storage_type<Component> *>(pools) == cpool ? nullptr : (other[pos] = std::get<storage_type<Component> *>(pools), other[pos++])), ...);
                                                                                                                                                              ^
/home/green/workspace/github/entt/src/entt/entity/storage.hpp:154:45: warning: Survived: Replaced - with + [cxx_sub_to_add]
            const auto pos = size_type(index-1u);
                                            ^
/home/green/workspace/github/entt/test/entt/entity/view.cpp:783:15: warning: Survived: Replaced 'a = b' with 'a = 42' [cxx_assign_const]
        check = false;
              ^
/home/green/workspace/github/entt/test/entt/entity/view.cpp:817:15: warning: Survived: Replaced 'a = b' with 'a = 42' [cxx_assign_const]
        check = false;
              ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with >= [cxx_lt_to_ge]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with <= [cxx_lt_to_le]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with >= [cxx_lt_to_ge]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with <= [cxx_lt_to_le]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with >= [cxx_lt_to_ge]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with <= [cxx_lt_to_le]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with >= [cxx_lt_to_ge]
            return lhs->size() < rhs->size();
                               ^
/home/green/workspace/github/entt/src/entt/entity/view.hpp:226:32: warning: Survived: Replaced < with <= [cxx_lt_to_le]
            return lhs->size() < rhs->size();
                               ^
[info] Mutation score: 61%
[info] Total execution time: 15006ms
```
</details>

--- a/include/mull/Driver.h
+++ b/include/mull/Driver.h
@@ -56,12 +56,15 @@ private:
   void selectInstructions(std::vector<FunctionUnderTest> &functions);
 
   std::vector<std::unique_ptr<MutationResult>>
-  runMutations(std::vector<MutationPoint *> &mutationPoints);
+  runMutations(std::vector<MutationPoint *> &mutationPoints,
+               std::vector<std::unique_ptr<Mutant>> &mutants);
 
   std::vector<std::unique_ptr<MutationResult>>
-  dryRunMutations(const std::vector<MutationPoint *> &mutationPoints);
+  dryRunMutations(const std::vector<MutationPoint *> &mutationPoints,
+                  std::vector<std::unique_ptr<Mutant>> &mutants);
   std::vector<std::unique_ptr<MutationResult>>
-  normalRunMutations(const std::vector<MutationPoint *> &mutationPoints);
+  normalRunMutations(const std::vector<MutationPoint *> &mutationPoints,
+                     std::vector<std::unique_ptr<Mutant>> &mutants);
 
   std::vector<FunctionUnderTest> getFunctionsUnderTest();
 };

--- a/include/mull/Mutant.h
+++ b/include/mull/Mutant.h
@@ -1,0 +1,34 @@
+#pragma once
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace mull {
+
+class MutationPoint;
+struct SourceLocation;
+
+class Mutant {
+public:
+  Mutant(std::string identifier, std::vector<MutationPoint *> points);
+
+  const std::string &getIdentifier() const;
+  const std::vector<MutationPoint *> &getMutationPoints() const;
+  const SourceLocation &getSourceLocation() const;
+  const std::string &getDiagnostics() const;
+  const std::string &getReplacement() const;
+  const std::string &getMutatorIdentifier() const;
+
+private:
+  std::string identifier;
+  std::string mutatorIdentifier;
+  std::vector<MutationPoint *> points;
+};
+
+struct MutantComparator {
+  bool operator()(std::unique_ptr<Mutant> &lhs, std::unique_ptr<Mutant> &rhs);
+  bool operator()(Mutant &lhs, Mutant &rhs);
+};
+
+} // namespace mull

--- a/include/mull/MutationResult.h
+++ b/include/mull/MutationResult.h
@@ -1,24 +1,27 @@
 #pragma once
 
-#include "ExecutionResult.h"
-#include "MutationPoint.h"
+#include "mull/ExecutionResult.h"
 #include <utility>
 
 namespace mull {
 
-class MutationResult {
-  ExecutionResult Result;
-  MutationPoint *MutPoint;
+class Mutant;
 
+class MutationResult {
 public:
-  MutationResult(ExecutionResult R, MutationPoint *MP) : Result(std::move(R)), MutPoint(MP) {}
+  MutationResult(ExecutionResult result, Mutant *mutant)
+      : result(std::move(result)), mutant(mutant) {}
 
   ExecutionResult &getExecutionResult() {
-    return Result;
+    return result;
   }
-  MutationPoint *getMutationPoint() {
-    return MutPoint;
+  Mutant *getMutant() {
+    return mutant;
   }
+
+private:
+  ExecutionResult result;
+  Mutant *mutant;
 };
 
 } // namespace mull

--- a/include/mull/Parallelization/Tasks/DryRunMutantExecutionTask.h
+++ b/include/mull/Parallelization/Tasks/DryRunMutantExecutionTask.h
@@ -1,18 +1,17 @@
 #pragma once
 
-#include <mull/MutationResult.h>
+#include "mull/Mutant.h"
+#include "mull/MutationResult.h"
 
 namespace mull {
 class progress_counter;
-class MutationPoint;
 
 class DryRunMutantExecutionTask {
 public:
-  using In = const std::vector<MutationPoint *>;
+  using In = const std::vector<std::unique_ptr<Mutant>>;
   using Out = std::vector<std::unique_ptr<MutationResult>>;
   using iterator = In::const_iterator;
 
-  void operator()(iterator begin, iterator end, Out &storage,
-                  progress_counter &counter);
+  void operator()(iterator begin, iterator end, Out &storage, progress_counter &counter);
 };
 } // namespace mull

--- a/include/mull/Parallelization/Tasks/MutantExecutionTask.h
+++ b/include/mull/Parallelization/Tasks/MutantExecutionTask.h
@@ -1,18 +1,17 @@
 #pragma once
 
+#include "mull/Mutant.h"
 #include "mull/MutationResult.h"
 
 namespace mull {
 
-class MutationPoint;
 class progress_counter;
 class Diagnostics;
-
 struct Configuration;
 
 class MutantExecutionTask {
 public:
-  using In = const std::vector<MutationPoint *>;
+  using In = const std::vector<std::unique_ptr<Mutant>>;
   using Out = std::vector<std::unique_ptr<MutationResult>>;
   using iterator = In::const_iterator;
 

--- a/include/mull/Result.h
+++ b/include/mull/Result.h
@@ -1,5 +1,6 @@
 #pragma once
 
+#include "Mutant.h"
 #include "MutationPoint.h"
 #include "MutationResult.h"
 #include <vector>
@@ -7,13 +8,16 @@
 namespace mull {
 
 class Result {
-  std::vector<std::unique_ptr<MutationResult>> mutationResults;
-  std::vector<MutationPoint *> mutationPoints;
-
 public:
-  Result(std::vector<std::unique_ptr<MutationResult>> mutationResults,
+  Result(std::vector<std::unique_ptr<Mutant>> mutants,
+         std::vector<std::unique_ptr<MutationResult>> mutationResults,
          std::vector<MutationPoint *> mutationPoints)
-      : mutationResults(std::move(mutationResults)), mutationPoints(std::move(mutationPoints)) {}
+      : mutants(std::move(mutants)), mutationResults(std::move(mutationResults)),
+        mutationPoints(std::move(mutationPoints)) {}
+
+  std::vector<std::unique_ptr<Mutant>> const &getMutants() const {
+    return mutants;
+  }
 
   std::vector<std::unique_ptr<MutationResult>> const &getMutationResults() const {
     return mutationResults;
@@ -22,5 +26,10 @@ public:
   std::vector<MutationPoint *> const &getMutationPoints() const {
     return mutationPoints;
   }
+
+private:
+  std::vector<std::unique_ptr<Mutant>> mutants;
+  std::vector<std::unique_ptr<MutationResult>> mutationResults;
+  std::vector<MutationPoint *> mutationPoints;
 };
 } // namespace mull

--- a/lib/CMakeLists.txt
+++ b/lib/CMakeLists.txt
@@ -14,6 +14,7 @@ set(mull_sources
   BitcodeLoader.cpp
   BitcodeMetadataReader.cpp
   MutationsFinder.cpp
+  Mutant.cpp
 
   Parallelization/Tasks/LoadBitcodeFromBinaryTask.cpp
 

--- a/lib/Driver.cpp
+++ b/lib/Driver.cpp
@@ -4,6 +4,7 @@
 #include "mull/Diagnostics/Diagnostics.h"
 #include "mull/Filters/Filters.h"
 #include "mull/Filters/FunctionFilter.h"
+#include "mull/Mutant.h"
 #include "mull/MutationResult.h"
 #include "mull/MutationsFinder.h"
 #include "mull/Parallelization/Parallelization.h"
@@ -34,9 +35,23 @@ Driver::~Driver() {
 std::unique_ptr<Result> Driver::run() {
   auto mutationPoints = findMutationPoints();
   auto filteredMutations = filterMutations(std::move(mutationPoints));
-  auto mutationResults = runMutations(filteredMutations);
 
-  return std::make_unique<Result>(std::move(mutationResults), std::move(filteredMutations));
+  std::vector<std::unique_ptr<Mutant>> mutants;
+  singleTask.execute("Deduplicate mutants", [&]() {
+    std::unordered_map<std::string, std::vector<MutationPoint *>> mapping;
+    for (MutationPoint *point : filteredMutations) {
+      mapping[point->getUserIdentifier()].push_back(point);
+    }
+    for (auto &pair : mapping) {
+      mutants.push_back(std::make_unique<Mutant>(pair.first, pair.second));
+    }
+    std::sort(std::begin(mutants), std::end(mutants), MutantComparator());
+  });
+
+  auto mutationResults = runMutations(filteredMutations, mutants);
+
+  return std::make_unique<Result>(
+      std::move(mutants), std::move(mutationResults), std::move(filteredMutations));
 }
 
 std::vector<MutationPoint *> Driver::findMutationPoints() {
@@ -129,22 +144,24 @@ void Driver::selectInstructions(std::vector<FunctionUnderTest> &functions) {
 }
 
 std::vector<std::unique_ptr<MutationResult>>
-Driver::runMutations(std::vector<MutationPoint *> &mutationPoints) {
+Driver::runMutations(std::vector<MutationPoint *> &mutationPoints,
+                     std::vector<std::unique_ptr<Mutant>> &mutants) {
   if (mutationPoints.empty()) {
     return std::vector<std::unique_ptr<MutationResult>>();
   }
 
   if (config.dryRunEnabled) {
-    return dryRunMutations(mutationPoints);
+    return dryRunMutations(mutationPoints, mutants);
   }
 
-  return normalRunMutations(mutationPoints);
+  return normalRunMutations(mutationPoints, mutants);
 }
 
 #pragma mark -
 
 std::vector<std::unique_ptr<MutationResult>>
-Driver::dryRunMutations(const std::vector<MutationPoint *> &mutationPoints) {
+Driver::dryRunMutations(const std::vector<MutationPoint *> &mutationPoints,
+                        std::vector<std::unique_ptr<Mutant>> &mutants) {
   std::vector<std::unique_ptr<MutationResult>> mutationResults;
 
   std::vector<DryRunMutantExecutionTask> tasks;
@@ -153,14 +170,15 @@ Driver::dryRunMutations(const std::vector<MutationPoint *> &mutationPoints) {
     tasks.emplace_back(DryRunMutantExecutionTask());
   }
   TaskExecutor<DryRunMutantExecutionTask> mutantRunner(
-      diagnostics, "Running mutants (dry run)", mutationPoints, mutationResults, std::move(tasks));
+      diagnostics, "Running mutants (dry run)", mutants, mutationResults, std::move(tasks));
   mutantRunner.execute();
 
   return mutationResults;
 }
 
 std::vector<std::unique_ptr<MutationResult>>
-Driver::normalRunMutations(const std::vector<MutationPoint *> &mutationPoints) {
+Driver::normalRunMutations(const std::vector<MutationPoint *> &mutationPoints,
+                           std::vector<std::unique_ptr<Mutant>> &mutants) {
   singleTask.execute("Prepare mutations", [&]() {
     for (auto point : mutationPoints) {
       point->getBitcode()->addMutation(point);
@@ -236,7 +254,7 @@ Driver::normalRunMutations(const std::vector<MutationPoint *> &mutationPoints) {
     tasks.emplace_back(config, diagnostics, executable, baseline);
   }
   TaskExecutor<MutantExecutionTask> mutantRunner(
-      diagnostics, "Running mutants", mutationPoints, mutationResults, std::move(tasks));
+      diagnostics, "Running mutants", mutants, mutationResults, std::move(tasks));
   mutantRunner.execute();
 
   return mutationResults;

--- a/lib/Mutant.cpp
+++ b/lib/Mutant.cpp
@@ -1,0 +1,44 @@
+#include "mull/Mutant.h"
+#include "mull/MutationPoint.h"
+#include "mull/SourceLocation.h"
+
+using namespace mull;
+
+Mutant::Mutant(std::string identifier, std::vector<MutationPoint *> points)
+    : identifier(std::move(identifier)), mutatorIdentifier(points.front()->getMutatorIdentifier()),
+      points(std::move(points)) {}
+
+const std::string &Mutant::getIdentifier() const {
+  return identifier;
+}
+
+const std::vector<MutationPoint *> &Mutant::getMutationPoints() const {
+  return points;
+}
+
+const SourceLocation &Mutant::getSourceLocation() const {
+  return points.front()->getSourceLocation();
+}
+
+const std::string &Mutant::getDiagnostics() const {
+  return points.front()->getDiagnostics();
+}
+
+const std::string &Mutant::getMutatorIdentifier() const {
+  return mutatorIdentifier;
+}
+
+const std::string &Mutant::getReplacement() const {
+  return points.front()->getReplacement();
+}
+
+bool MutantComparator::operator()(std::unique_ptr<Mutant> &lhs, std::unique_ptr<Mutant> &rhs) {
+  return operator()(*lhs, *rhs);
+}
+
+bool MutantComparator::operator()(Mutant &lhs, Mutant &rhs) {
+  const SourceLocation &l = lhs.getSourceLocation();
+  const SourceLocation &r = rhs.getSourceLocation();
+  return std::tie(l.filePath, l.line, l.column, lhs.getMutatorIdentifier()) <
+         std::tie(r.filePath, r.line, r.column, rhs.getMutatorIdentifier());
+}

--- a/lib/Parallelization/Tasks/DryRunMutantExecutionTask.cpp
+++ b/lib/Parallelization/Tasks/DryRunMutantExecutionTask.cpp
@@ -1,5 +1,4 @@
 #include "mull/Parallelization/Tasks/DryRunMutantExecutionTask.h"
-
 #include "mull/Parallelization/Progress.h"
 
 using namespace mull;
@@ -7,9 +6,9 @@ using namespace mull;
 void DryRunMutantExecutionTask::operator()(iterator begin, iterator end, Out &storage,
                                            progress_counter &counter) {
   for (auto it = begin; it != end; it++, counter.increment()) {
-    auto mutationPoint = *it;
+    auto &mutant = *it;
     ExecutionResult result;
     result.status = DryRun;
-    storage.push_back(std::make_unique<MutationResult>(result, mutationPoint));
+    storage.push_back(std::make_unique<MutationResult>(result, mutant.get()));
   }
 }

--- a/lib/Parallelization/Tasks/MutantExecutionTask.cpp
+++ b/lib/Parallelization/Tasks/MutantExecutionTask.cpp
@@ -18,13 +18,13 @@ void MutantExecutionTask::operator()(iterator begin, iterator end, Out &storage,
                                      progress_counter &counter) {
   Runner runner(diagnostics);
   for (auto it = begin; it != end; ++it, counter.increment()) {
-    MutationPoint *mutationPoint = *it;
+    auto &mutant = *it;
     ExecutionResult result = runner.runProgram(executable,
                                                {},
-                                               { mutationPoint->getUserIdentifier() },
+                                               { mutant->getIdentifier() },
                                                baseline.runningTime * 10,
                                                configuration.captureMutantOutput);
 
-    storage.push_back(std::make_unique<MutationResult>(result, mutationPoint));
+    storage.push_back(std::make_unique<MutationResult>(result, mutant.get()));
   }
 }

--- a/lib/Reporters/IDEReporter.cpp
+++ b/lib/Reporters/IDEReporter.cpp
@@ -1,6 +1,7 @@
 #include "mull/Reporters/IDEReporter.h"
 
 #include "mull/Diagnostics/Diagnostics.h"
+#include "mull/Mutant.h"
 #include "mull/MutationResult.h"
 #include "mull/Mutators/Mutator.h"
 #include "mull/Reporters/SourceCodeReader.h"
@@ -19,7 +20,7 @@ static bool mutantSurvived(const ExecutionStatus &status) {
 }
 
 static void printMutant(Diagnostics &diagnostics, SourceCodeReader &sourceCodeReader,
-                        const MutationPoint &mutant, bool survived) {
+                        const Mutant &mutant, bool survived) {
   auto &sourceLocation = mutant.getSourceLocation();
   assert(!sourceLocation.isNull() && "Debug information is missing?");
 
@@ -39,14 +40,14 @@ IDEReporter::IDEReporter(Diagnostics &diagnostics, bool showKilled)
     : diagnostics(diagnostics), showKilled(showKilled), sourceCodeReader() {}
 
 void IDEReporter::reportResults(const Result &result) {
-  if (result.getMutationPoints().empty()) {
+  if (result.getMutants().empty()) {
     diagnostics.info("No mutants found. Mutation score: infinitely high");
     return;
   }
 
-  std::set<MutationPoint *> killedMutants;
+  std::set<Mutant *> killedMutants;
   for (auto &mutationResult : result.getMutationResults()) {
-    auto mutant = mutationResult->getMutationPoint();
+    auto mutant = mutationResult->getMutant();
     auto &executionResult = mutationResult->getExecutionResult();
 
     if (!mutantSurvived(executionResult.status)) {
@@ -54,7 +55,7 @@ void IDEReporter::reportResults(const Result &result) {
     }
   }
 
-  auto survivedMutantsCount = result.getMutationPoints().size() - killedMutants.size();
+  auto survivedMutantsCount = result.getMutants().size() - killedMutants.size();
   auto killedMutantsCount = killedMutants.size();
 
   /// It is important that below we iterate over result.getMutationPoints()
@@ -62,11 +63,11 @@ void IDEReporter::reportResults(const Result &result) {
   /// Optimizing this here was a bad idea: https://github.com/mull-project/mull/pull/640.
   if (showKilled && killedMutantsCount > 0) {
     std::stringstream stringstream;
-    stringstream << "Killed mutants (" << killedMutantsCount << "/"
-                 << result.getMutationPoints().size() << "):";
+    stringstream << "Killed mutants (" << killedMutantsCount << "/" << result.getMutants().size()
+                 << "):";
     diagnostics.info(stringstream.str());
-    for (auto &mutant : result.getMutationPoints()) {
-      if (killedMutants.find(mutant) != killedMutants.end()) {
+    for (auto &mutant : result.getMutants()) {
+      if (killedMutants.find(mutant.get()) != killedMutants.end()) {
         printMutant(diagnostics, sourceCodeReader, *mutant, false);
       }
     }
@@ -75,11 +76,11 @@ void IDEReporter::reportResults(const Result &result) {
   if (survivedMutantsCount > 0) {
     std::stringstream stringstream;
     stringstream << "Survived mutants (" << survivedMutantsCount << "/"
-                 << result.getMutationPoints().size() << "):";
+                 << result.getMutants().size() << "):";
     diagnostics.info(stringstream.str());
 
-    for (auto &mutant : result.getMutationPoints()) {
-      if (killedMutants.find(mutant) == killedMutants.end()) {
+    for (auto &mutant : result.getMutants()) {
+      if (killedMutants.find(mutant.get()) == killedMutants.end()) {
         printMutant(diagnostics, sourceCodeReader, *mutant, true);
       }
     }
@@ -87,7 +88,7 @@ void IDEReporter::reportResults(const Result &result) {
     diagnostics.info("All mutations have been killed");
   }
 
-  auto rawScore = double(killedMutants.size()) / double(result.getMutationPoints().size());
+  auto rawScore = double(killedMutants.size()) / double(result.getMutants().size());
   auto score = uint(rawScore * 100);
   diagnostics.info(std::string("Mutation score: ") + std::to_string(score) + '%');
 }

--- a/lib/Reporters/MutationTestingElementsReporter.cpp
+++ b/lib/Reporters/MutationTestingElementsReporter.cpp
@@ -1,6 +1,7 @@
 #include "mull/Reporters/MutationTestingElementsReporter.h"
 
 #include "mull/Diagnostics/Diagnostics.h"
+#include "mull/Mutant.h"
 #include "mull/MutationResult.h"
 #include "mull/Mutators/Mutator.h"
 #include "mull/Reporters/ASTSourceInfoProvider.h"
@@ -25,7 +26,7 @@ static bool mutantSurvived(const ExecutionStatus &status) {
 }
 
 static json11::Json createFiles(Diagnostics &diagnostics, const Result &result,
-                                const std::set<MutationPoint *> &killedMutants,
+                                const std::set<Mutant *> &killedMutants,
                                 SourceInfoProvider &sourceInfoProvider) {
   SourceManager sourceManager;
 
@@ -34,13 +35,13 @@ static json11::Json createFiles(Diagnostics &diagnostics, const Result &result,
   // Mutation Testing Elements Schema suggests storing mutation points based
   // on their source file.
   // Step 1: we create a map: "file => all of its mutation points".
-  std::map<std::string, std::vector<MutationPoint *>> mutationPointsPerFile;
-  for (auto mutant : result.getMutationPoints()) {
+  std::map<std::string, std::vector<Mutant *>> mutationPointsPerFile;
+  for (auto &mutant : result.getMutants()) {
     auto &sourceLocation = mutant->getSourceLocation();
     auto sourceCodeLine = sourceManager.getLine(sourceLocation);
     assert(sourceLocation.column < sourceCodeLine.size());
 
-    mutationPointsPerFile[sourceLocation.filePath].push_back(mutant);
+    mutationPointsPerFile[sourceLocation.filePath].push_back(mutant.get());
   }
 
   // Step 2: Iterate through each file and dump the information about each
@@ -57,16 +58,16 @@ static json11::Json createFiles(Diagnostics &diagnostics, const Result &result,
 
     Json::array mutantsEntries;
 
-    for (MutationPoint *mutationPoint : fileMutationPoints.second) {
+    for (Mutant *mutant : fileMutationPoints.second) {
       MutationPointSourceInfo sourceInfo =
-          sourceInfoProvider.getSourceInfo(diagnostics, mutationPoint);
+          sourceInfoProvider.getSourceInfo(diagnostics, mutant->getMutationPoints().front());
 
-      std::string status = (killedMutants.count(mutationPoint) == 0) ? "Survived" : "Killed";
+      std::string status = (killedMutants.count(mutant) == 0) ? "Survived" : "Killed";
 
       Json mpJson =
-          Json::object{ { "id", mutationPoint->getMutator()->getUniqueIdentifier() },
-                        { "mutatorName", mutationPoint->getDiagnostics() },
-                        { "replacement", mutationPoint->getReplacement() },
+          Json::object{ { "id", mutant->getMutatorIdentifier() },
+                        { "mutatorName", mutant->getDiagnostics() },
+                        { "replacement", mutant->getReplacement() },
                         { "location",
                           Json::object{ { "start",
                                           Json::object{ { "line", sourceInfo.beginLine },
@@ -118,9 +119,9 @@ void MutationTestingElementsReporter::reportResults(const Result &result) {
   }
   generateHTMLFile();
 
-  std::set<MutationPoint *> killedMutants;
+  std::set<Mutant *> killedMutants;
   for (auto &mutationResult : result.getMutationResults()) {
-    auto mutant = mutationResult->getMutationPoint();
+    auto mutant = mutationResult->getMutant();
     auto &executionResult = mutationResult->getExecutionResult();
 
     if (!mutantSurvived(executionResult.status)) {

--- a/lib/Reporters/SQLiteReporter.cpp
+++ b/lib/Reporters/SQLiteReporter.cpp
@@ -3,6 +3,7 @@
 #include "mull/Bitcode.h"
 #include "mull/Diagnostics/Diagnostics.h"
 #include "mull/ExecutionResult.h"
+#include "mull/Mutant.h"
 #include "mull/MutationResult.h"
 #include "mull/Mutators/Mutator.h"
 #include "mull/Result.h"
@@ -85,8 +86,7 @@ void mull::SQLiteReporter::reportResults(const Result &result) {
   sqlite3_prepare(database, insertExecutionResultQuery, -1, &insertExecutionResultStmt, nullptr);
 
   for (auto &mutationResult : result.getMutationResults()) {
-    MutationPoint *mutationPoint = mutationResult->getMutationPoint();
-    const std::string &pointId = mutationPoint->getUserIdentifier();
+    const std::string &pointId = mutationResult->getMutant()->getIdentifier();
 
     ExecutionResult mutationExecutionResult = mutationResult->getExecutionResult();
 
@@ -114,8 +114,7 @@ void mull::SQLiteReporter::reportResults(const Result &result) {
   }
 
   const char *insertMutationPointQuery =
-      "INSERT OR IGNORE INTO mutation_point VALUES (?1, ?2, ?3, ?4, ?5, ?6, "
-      "?7, ?8, ?9, ?10, ?11, ?12)";
+      "INSERT INTO mutation_point VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7, ?8, ?9, ?10, ?11, ?12)";
   sqlite3_stmt *insertMutationPointStmt;
   sqlite3_prepare(database, insertMutationPointQuery, -1, &insertMutationPointStmt, nullptr);
 
@@ -197,7 +196,7 @@ CREATE TABLE mutation_point (
   ideDiagnostics TEXT,
   line_number INT,
   column_number INT,
-  unique_id TEXT UNIQUE
+  unique_id TEXT
 );
 )CreateTables";
 

--- a/tests-lit/tests/deduplication/01/main.cpp
+++ b/tests-lit/tests/deduplication/01/main.cpp
@@ -1,0 +1,25 @@
+template <typename T> T doSomething(T arg1, T arg2) {
+  return arg1 + arg2;
+}
+
+int main() {
+  if (doSomething(2, 3) != 5) {
+    return 1;
+  }
+  if (doSomething(2.0, 0.0) != 2.0) {
+    return 1;
+  }
+  return 0;
+}
+
+// clang-format off
+
+// RUN: cd / && %clang_cxx %s -fembed-bitcode -g -o %s.exe
+// RUN: cd %CURRENT_DIR
+
+// RUN: unset TERM; %MULL_EXEC -linker=%clang_cxx -mutators=cxx_add_to_sub -ide-reporter-show-killed %s.exe 2>&1 | %FILECHECK_EXEC %s --dump-input=fail --strict-whitespace --match-full-lines
+// CHECK:[info] Killed mutants (1/1):
+// CHECK:{{.*}}main.cpp:2:15: warning: Killed: Replaced + with - [cxx_add_to_sub]
+// CHECK:  return arg1 + arg2;
+// CHECK:              ^
+// CHECK:[info] Mutation score: 100%

--- a/tests/MutationTestingElementsReporterTest.cpp
+++ b/tests/MutationTestingElementsReporterTest.cpp
@@ -96,14 +96,18 @@ TEST(MutationTestingElementsReporterTest, integrationTest) {
   mutatedTestExecutionResult.stdoutOutput = "mutatedTestExecutionResult.STDOUT";
   mutatedTestExecutionResult.stderrOutput = "mutatedTestExecutionResult.STDERR";
 
-  auto mutationResult = std::make_unique<MutationResult>(mutatedTestExecutionResult, mutationPoint);
+  auto mutant = std::make_unique<Mutant>(mutationPoint->getUserIdentifier(), mutationPoints);
+  auto mutationResult = std::make_unique<MutationResult>(mutatedTestExecutionResult, mutant.get());
+
+  std::vector<std::unique_ptr<Mutant>> mutants;
+  mutants.push_back(std::move(mutant));
 
   std::vector<std::unique_ptr<MutationResult>> mutationResults;
   mutationResults.push_back(std::move(mutationResult));
 
   MetricsMeasure resultTime;
 
-  Result result(std::move(mutationResults), mutationPoints);
+  Result result(std::move(mutants), std::move(mutationResults), mutationPoints);
 
   /// STEP2. Reporting results to JSON
   MockASTSourceInfoProvider sourceInfoProvider;

--- a/tests/SQLiteReporterTest.cpp
+++ b/tests/SQLiteReporterTest.cpp
@@ -4,6 +4,7 @@
 #include "mull/BitcodeLoader.h"
 #include "mull/Config/Configuration.h"
 #include "mull/Metrics/MetricsMeasure.h"
+#include "mull/Mutant.h"
 #include "mull/MutationsFinder.h"
 #include "mull/Program/Program.h"
 #include "mull/ReachableFunction.h"
@@ -68,14 +69,18 @@ TEST(SQLiteReporter, integrationTest) {
   mutatedTestExecutionResult.stdoutOutput = "mutatedTestExecutionResult.STDOUT";
   mutatedTestExecutionResult.stderrOutput = "mutatedTestExecutionResult.STDERR";
 
-  auto mutationResult = std::make_unique<MutationResult>(mutatedTestExecutionResult, mutationPoint);
+  auto mutant = std::make_unique<Mutant>(mutationPoint->getUserIdentifier(), mutationPoints);
+  auto mutationResult = std::make_unique<MutationResult>(mutatedTestExecutionResult, mutant.get());
+
+  std::vector<std::unique_ptr<Mutant>> mutants;
+  mutants.push_back(std::move(mutant));
 
   std::vector<std::unique_ptr<MutationResult>> mutationResults;
   mutationResults.push_back(std::move(mutationResult));
 
   MetricsMeasure resultTime;
 
-  Result result(std::move(mutationResults), mutationPoints);
+  Result result(std::move(mutants), std::move(mutationResults), mutationPoints);
 
   /// STEP2. Reporting results to SQLite
   SQLiteReporter reporter(diagnostics, "integration test", "");

